modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/dht/topology/PartitionEvictionOrderTest.java

```diff
@@ -0,0 +1,199 @@
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(igniteInstanceName);
+
+        cfg.setDataStorageConfiguration(new DataStorageConfiguration()
+            .setDefaultDataRegionConfiguration(new DataRegionConfiguration().setPersistenceEnabled(true)));
+
+        CacheConfiguration<Long, Long> atomicCcfg = new CacheConfiguration<Long, Long>(DEFAULT_CACHE_NAME)
+            .setAtomicityMode(ATOMIC)
+            .setRebalanceMode(CacheRebalanceMode.ASYNC)
+            .setCacheMode(REPLICATED);
+
+        cfg.setCacheConfiguration(atomicCcfg);
+
+        return cfg;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /**
+     * Tests that {@link CacheRebalanceMode#SYNC} caches are evicted at first.
+     */
+    @Test
+    @WithSystemProperty(key = IgniteSystemProperties.IGNITE_EVICTION_PERMITS, value = "1")
+    public void testSyncCachesEvictedAtFirst() throws Exception {
+        IgniteEx node0 = startGrid(0);
+
+        node0.cluster().active(true);
+
+        IgniteEx node1 = startGrid(1);
+
+        node0.cluster().setBaselineTopology(node1.cluster().topologyVersion());
+
+        GridCacheAdapter<Object, Object> utilCache0 = grid(0).context().cache().internalCache(CU.UTILITY_CACHE_NAME);
+
+        IgniteCache<Object, Object> cache = node0.getOrCreateCache(DEFAULT_CACHE_NAME);
+
+        for (int i = 0; i < 100_000; i++) {
+            utilCache0.put(i, i);
+
+            cache.put(i, i);
+        }
+
+        awaitPartitionMapExchange();
+
+        stopGrid(0);
+
+        GridCacheAdapter<Object, Object> utilCache1 = grid(1).context().cache().internalCache(CU.UTILITY_CACHE_NAME);
+
+        IgniteCache<Object, Object> cache2 = grid(1).cache(DEFAULT_CACHE_NAME);
+
+        for (int i = 0; i < 200_000; i++) {
+            try {
+                cache2.put(i, i + 1);
+
+                utilCache1.put(i, i + 1);
+            }
+            catch (IgniteCheckedException e) {
+                e.printStackTrace();
+            }
+        }
+
+        CountDownLatch latch = new CountDownLatch(1);
+
+        TestDependencyResolver resolver = new TestDependencyResolver(new DependencyResolver() {
+            @Override public <T> T resolve(T instance) {
+                if (instance instanceof PartitionsEvictManager.BucketQueue)
+                    return (T)(new TestBucketQueue(1, latch));
+
+                return instance;
+            }
+        });
+
+        startGrid(0, resolver);
+
+        //We start not from 0 because the first partition that comes to queue is from async cache.
+        for (int i = 1; i < 101; i++)
```

Comment:
This check looks too strong. Is there a guarantee that first partition will be from ASYNC cache?
By the way, why first partition should be from ASYNC cache, if we wait for 5 seconds before first poll?